// ask-cli/lib/utils/isp-utils/isp-constants.js
'use strict';

module.exports = {
    /**
     * Common errors of ISP commands
     */
    ERROR: Object.freeze({
        INVALID_API_PARAMETERS: 'Invalid API parameters',
        INVALID_ISP_TYPE: 'Invalid in-skill product type',
        INVALID_ISP_STAGE: 'Invalid in-skill product stage',
        INVALID_ISP_STATUS: 'Invalid in-skill product status',
        INVALID_ISP_SKILL_ASSOCIATION: 'Invalid in-skill product skill association',
        INVALID_ISP_DEFINITION: 'Invalid in-skill product definition',
        INVALID_ISP_CONFIG: 'Invalid in-skill product configuration',
        INVALID_SKILL_STAGE: 'Invalid skill stage',
        INVALID_ISP_FILE: 'Invalid in-skill product file',
        INVALID_ISP_NAME: 'Invalid in-skill product name.\n  The valid in-skill product name can only contain letters (a-z), numbers (0-9), underscores (_) or dashes (-).\n  The minimum length of name is 3 and the maximum is 50.',
        INVALID_CONFIG_MISSING_FILE: 'Missing file',
        INVALID_CONFIG_MISSING_ID: 'Missing ID',
        INVALID_CONFIG_MISSING_ETAG: 'Missing eTag',
        INVALID_CONFIG_INVALID_STATUS: 'Invalid deploy status',
        INVALID_ISP_PURCHASABLE_STATE: 'Invalid purchasable state',
        MISSING_ISP_DEFINITION: 'Missing in-skill product definition',
        MISSING_VENDOR_ID: 'Missing vendor ID',
        MISSING_ISP_PRODUCT_ID: 'Missing in-skill product ID',
        MISSING_ISP_STAGE: 'Missing in-skill product stage',
        MISSING_SKILL_ID: 'Missing Skill ID',
        MISSING_ISP_TYPE: 'Missing in-skill product type',
        MISSING_ISP_TITLE: 'Missing in-skill product title',
        MISSING_ISP_FILE: 'Missing in-skill product file',
        MISSING_ISP_STATUS: 'Missing in-skill product status',
        MISSING_USER_PROFILE: 'Missing user profile information',
        MISSING_ISP_SKILL_ASSOCIATION: 'Missing in-skill product skill association',
        MISSING_SKILL_STAGE: 'Missing skill stage',
        MISMATCH_ETAG: 'The local stored in-skill product eTag does not match with the one in the server side.\n  Use "ask diff --target isp" to inspect the difference between local and remote.\n  Use "ask deploy --target isp --force" to deploy local in-skill product(s) regardless of eTag checking.',

        DEVELOPMENT_STAGE_ONLY: 'Only development stage supported',
        CANNOT_RETRIEVE_TEMPLATE_LIST_FROM_S3: 'Cannot retrieve in-skill product template list from S3 bucket',
        CANNOT_RETRIEVE_TEMPLATE: 'Cannot retrieve in-skill product template from the URL',
        CANNOT_GET_VENDOR_ID_FROM_PROFILE: 'Cannot get vendor id associated with the current profile',
        NO_TEMPLATE_ASSOCIATE: 'No template associate with the input name',
        NOT_AT_SKILL_PROJECT_ROOT: 'Not an ASK CLI skill project (Please execute this command at the root directory of an ASK CLI skill project or make sure an ASK CLI skill project has been created).',

        FAILED_TO_UPDATE_ISP_CONFIG: 'Failed to update in-skill product deploy setting',
        FAILED_TO_GET_ISP_INFO: 'Failed to get information for the given in-skill product',
        FAILED_TO_CLONE_ISP: 'Failed to clone in-skill products',
        FAILED_TO_GET_ISP_LIST_FOR_SKILL: 'Failed to get the list of in-skill products associated with given skill',
        FAILED_TO_GET_ISP_LIST_FOR_VENDOR: 'Failed to get the list of in-skill products associated with given vendor',
        FAILED_TO_GET_SKILL_LIST_FOR_ISP: 'Failed to get the list of skills associated with given in-skill product',
        FAILED_TO_CREATE_ISP: 'Failed to create in-skill product',
        FAILED_TO_UPDATE_ISP: 'Failed to update in-skill product',
        FAILED_TO_ASSOCIATE_ISP_TO_SKILL: 'Failed to associate in-skill product to a skill',
        FAILED_TO_DISASSOCIATE_ISP_FROM_SKILL: 'Failed to disassociate in-skill product from a skill',
        FAILED_TO_DELETE_ISP: 'Failed to delete in-skill product',
        FAILED_TO_RESET_ISP_ENTITLEMENT: 'Failed to reset in-skill product entitlement'
    }),

    /**
     * Common warnings of ISP commands
     */
    WARNING: Object.freeze({
        ISP_FILE_NOT_FOUND: 'Please verify in-skill product definition is in current working directory.',
        CLI_CONFIG_NOT_FOUND: 'Please make sure ~/.ask/cli_config exists.'
    }),

    URL: Object.freeze({
        ISP_BASE: '/inSkillProducts',
        SKILL_BASE: '/skills',
        S3_TEMPLATES_ENDPOINT: 'https://s3.amazonaws.com/ask-cli/isp/isp_template_index.json'
    }),

    DIRECTORY: Object.freeze({
        ISP_ROOT: 'isps',
        SUBSCRIPTION: 'subscription',
        ENTITLEMENT: 'entitlement',
        CONSUMABLE: 'consumable'
    }),

    ISP_STAGE: Object.freeze({
        LIVE: 'live',
        DEVELOPMENT: 'development'
    }),

    ISP_PRODUCT_TYPE: Object.freeze({
        SUBSCRIPTION: 'SUBSCRIPTION',
        ENTITLEMENT: 'ENTITLEMENT'
    }),

    ISP_STATUS: Object.freeze({
        INCOMPLETE: 'INCOMPLETE',
        COMPLETE: 'COMPLETE',
        CERTIFICATION: 'CERTIFICATION',
        PUBLISHED: 'PUBLISHED',
        SUPPRESSED: 'SUPPRESSED'
    }),

    ISP_PURCHASABLE_STATE: Object.freeze({
        PURCHASABLE: 'PURCHASABLE',
        NOT_PURCHASABLE: 'NOT_PURCHASABLE'
    }),

    SKILL_ASSOCIATION: Object.freeze({
        ASSOCIATED_WITH_SKILL: 'ASSOCIATED_WITH_SKILL',
        NOT_SKILL_ASSOCIATION: 'NOT_SKILL_ASSOCIATION'
    }),

    NUMBERS: Object.freeze({
        DEFAULT_ISP_MAX_RESULTS: 50,
        DEFAULT_ISP_MAX_ASYNC_REQUESTS: 1,
        MAX_ISP_NAME_LENGTH: 50,
        MIN_ISP_NAME_LENGTH: 3,
        TIME_INTERVAL_FOR_ASYNC_REQUEST: 1000
    }),

    API: Object.freeze({
        CREATE_ISP: 'create-isp',
        UPDATE_ISP: 'update-isp',
        GET_ISP: 'get-isp',
        ASSOCIATE_ISP: 'associate-isp',
        DISASSOCIATE_ISP: 'disassociate-isp',
        DELETE_ISP: 'delete-isp',
        LIST_SKILLS_FOR_ISP: 'list-skills-for-isp',
        LIST_ISP_FOR_SKILL: 'list-isp-for-skill',
        LIST_ISP_FOR_VENDOR: 'list-isp-for-vendor',
        RESET_ISP_ENTITLEMENT: 'reset-isp-entitlement'
    }),

    DEPLOY_STATUS: Object.freeze({
        ASSOCIATE: 'Associate',
        UPDATE: 'Update',
        ADD: 'Add',
        REMOVE: 'Remove'
    }),

    DEPLOY_STATUS_TABLE_HEADERS: Object.freeze(['DEPLOY STATUS', 'ID', 'FILE']),
    DEPLOY_ERROR_TABLE_HEADERS: Object.freeze(['ERROR', 'ID', 'FILE']),
    COMMAND: Object.freeze({
        ADD_ISP: {
            HELP_DESCRIPTION:
            '\n  Usage: ask-add-isp\n\n' +
            '  $ ask add isp [-i|--isp-id <isp-id>] [-f|--file <file-path>] [--isp-name <name>]\n\n' +
            '  Add an in-skill product to the skill project by template(run without option), file([--file <file-path>]) or ID([--isp-id <isp-id>],\n' +
            '  the added in-skill product will be created and/or associate to the skill during deployment("ask deploy [--target isp]" is run).\n' +
            '  Reset an in-skill product\'s deploy status from  "Remove" to "Update".\n\n' +
            '    --isp-id <isp-id>              add an existing in-skill product to current skill project by given ID\n' +
            '    -f, --file <file-path>         add an in-skill product to current skill project by given the path of the definition file\n' +
            '    --isp-name <name>              assign a name to the new in-skill product when using template\n' +
            '                                   a valid in-skill product name should have at least 3 characters and at most 50 characters\n'+
            '                                   that contains only letters (a-z), numbers (0-9), underscores (_) or dashes (-).\n' +
            '    -p, --profile <profile>        add a component under the chosen profile\n' +
            '    -h, --help                     output usage information'
        },
        REMOVE_ISP: {
            HELP_DESCRIPTION:
            '\n  Usage: ask-remove-isp\n\n' +
            '  $ ask remove isp [-i|--isp-id <isp-id>] [-f|--file <file-path>]\n\n' +
            '  Remove an in-skill product from current skill by file([--file <file-path>]) or ID ([--isp-id <isp-id>])\n'+
            '  if it is currently associated with the skill, the in-skill product\'s deploy status is set to "Remove"\n'+
            '  so it will be disassociated from skill during deployment("ask deploy [--target isp]" is run).\n' +
            '  Exclude an in-skill product from deployment if its deploy status is  "Add" or "Associate".\n\n' +
            '    --isp-id <isp-id>              remove an existing in-skill product to current skill project by given in-skill product ID\n' +
            '    -f, --file                     remove an in-skill product from the current skill project by given path of the definition file\n' +
            '    -p, --profile <profile>        remove a component under the chosen profile\n' +
            '    -h, --help                     output usage information'
        }
    })
};
