// ask-cli/lib/init/init.js
'use strict';

const initHomeConfig = require('./init-home-config');
const setVendor = require('./set-vendor');
const awsResolver = require('./aws-resolver');
const inquirer = require('inquirer');
const profileHelper = require('../utils/profile-helper');
const CONSTANTS = require('../utils/constants');
const awsSetup = require('../utils/aws-setup');
const lwaUtil = require('../utils/lwa');
const oauthWrapper = require('../utils/oauth-wrapper');

const LIST_PAGE_SIZE = 50;


module.exports = {
    createCommand: (commander) => {
        commander
            .command('init')
            .description('initialize the ask-cli with your Amazon developer account credentials')
            .option('--no-browser', 'display authorization url instead of opening browser')
            .option('-l, --list-profiles', 'list all the profile(s) for ask-cli')
            .option('-p, --profile <profile>', 'name for the profile to be created/updated')
            .option('--debug', 'ask cli debug mode')
            .option('--aws-setup', 'set up AWS profile with "Access Key ID" and "Secret Access Key"')
            .option('-h, --help', 'output usage information', () => {
                console.log(CONSTANTS.COMMAND.INIT.HELP_DESCRIPTION);
                process.exit(0);
            })
            .action(handle);

        function handle(options) {
            if (options && typeof options === 'string') {
                console.error('[Error]: Invalid command. Please run "ask init -h" for help.');
                return;
            }
            if (options.listProfiles) {
                profileHelper.displayProfile();
                return;
            }
            if (options.awsSetup) {
                awsSetup.setup((error, profileName) => {
                    if (error) {
                        console.error(error);
                    } else {
                        console.log('AWS profile, "' + profileName + '", has been set.');
                    }
                });
                return;
            }
            let firstTimeCreation = initHomeConfig.initHomeConfig();            
            if (options.profile || firstTimeCreation) {
                let profile = options.profile || 'default';
                directInitProcess(options.browser, profile, options.debug);
            } else {
                listInitProcess(options.browser, options.debug);
            }
        }
    }
};

function directInitProcess(browser, askProfile, doDebug) {
    // the check below is necessary, since there's a case that a user can directly name a profile
    // as `ask init -p 'something something'`. It can bypass the validation inside the 'newProfile' function.
    if (profileHelper.askProfileSyntaxValidation(askProfile) === false) {
        console.error('[Error]: Illegal profile name. The profile name has to match with the pattern: /[a-zA-Z0-9-_]/');
        process.exit(1);
    }
    console.log('-------------------- Initialize CLI --------------------');
    awsResolver.resolve(askProfile, (error) => {
        if (error) {
            console.error(error);
            process.exit(1);
        }
        lwaUtil.accessTokenGenerator(CONSTANTS.LWA.CLI_DEFAULT_CREDENTIALS,
            CONSTANTS.LWA.DEFAULT_SCOPES,
            CONSTANTS.LWA.DEFAULT_STATE,
            browser,
            (error, tokens) => {
                if (error) {
                    // if there's an error, delete the
                    // half finished askProfile created from last step
                    profileHelper.deleteProfile(askProfile);
                    console.error('[Error]: ' + error);
                    process.exit(1);
                } else {
                    oauthWrapper.writeToken(tokens, askProfile);
                    console.log('Tokens fetched and recorded in ask-cli config.');
                    setVendor.setVendorId(askProfile, doDebug);
                }
            }
        );
    });
}

function listInitProcess(browser, doDebug) {
    let rawProfileList = profileHelper.getListProfile();
    if (!rawProfileList || rawProfileList.length === 0) {
        newProfile(browser);
        return;
    }
    const HEADER = 'Profile              Associated AWS Profile';
    const CREATE_PROFILE_MESSAGE = 'Create new profile';
    let profileList = profileHelper.stringFormatter(profileHelper.getListProfile());
    if (!profileList || profileList.length === 0) {
        newProfile(browser);
    } else {
        profileList.splice(0, 0, new inquirer.Separator());
        profileList.splice(1, 0, CREATE_PROFILE_MESSAGE);
        profileList.splice(2, 0, new inquirer.Separator());
        profileList.splice(3, 0, new inquirer.Separator(HEADER));
        inquirer.prompt([{
            type: 'list',
            message: 'Please create a new profile or overwrite the existing profile.\n',
            name: 'profile',
            pageSize: LIST_PAGE_SIZE,
            choices: profileList
        }]).then((answers) => {
            if (answers.profile === CREATE_PROFILE_MESSAGE) {
                newProfile(browser);
            } else {
                let overrideProfile = answers.profile.split(' ')[0].trim();
                let profileName = overrideProfile.match(/^\[([a-zA-Z0-9-_]+)\]$/);
                if (!profileName || !profileName[1]) {
                    console.log('[Error]: Profile name invalid. Please make sure the selected ' + 
                        'profile name follows /[a-zA-Z0-9-_]/ and doesn\'t include spaces.');
                } else {
                    directInitProcess(browser, profileName[1], doDebug);
                }
            }
        });
    }
}

function newProfile(browser) {
    inquirer.prompt([{
        message: 'Please type in your new profile name:\n',
        type: 'input',
        name: 'profile',
        default: 'default'
    }]).then((answer) => {
        let newProfileName = answer.profile.trim();
        if (!profileHelper.askProfileSyntaxValidation(newProfileName)) {
            console.error('[Error]: Invalid profile name. The profile name has to match with the pattern: /[a-zA-Z0-9-_]/');
            process.exit(1);
        }
        directInitProcess(browser, answer.profile.trim());
    });
}
